sklearn/metrics/pairwise.py

```diff
@@ -674,6 +689,203 @@ def cosine_distances(X, Y=None):
+                                 force_all_finite=False)
+
+    X = np.asarray(X, dtype=np.object)
+
+    cat_mask, num_mask = \
+        _detect_categorical_features(X, categorical_features)
+
+    # Calculates the min and max values, and if requested, scale the
+    # input values in order to obtain the distances between 0 and 1,
+    # as proposed by the Gower's paper.
+    ranges = 1
+    if np.any(num_mask):
+        process_scale = False
+        if isinstance(scale, bool):
+            process_scale = scale
+        else:
+            if len(np.asarray(scale).flatten()) != X[:, num_mask].shape[1]:
+                raise ValueError("Length of scale parameter must be equal "
+                                 "to the number of numerical columns.")
+            process_scale = True
+
+        ranges, min, max = _precompute_gower_params(X, Y, scale, num_mask)
+
+        # avoid division by zero when all values in the column are the same
+        ranges[ranges == 0] = 1
+
+        # check if the data is pre-scaled when scale=False
+        if not process_scale and (np.min(min) < 0 or np.max(max) > 1):
+            raise ValueError("Input data is not scaled between 0 and 1.")
+
+    D = np.zeros((X.shape[0], Y.shape[0]), dtype=np.float)
+
+    for i in range(X.shape[0]):
+        j_start = i
+
+        # For non square results
+        if X.shape[0] != Y.shape[0] or X is not Y:
+            j_start = 0
+
+        # Calculates the similarities for categorical columns
+        cat_dists = (X[i, cat_mask] != Y[j_start:, cat_mask]) | \
+                    (_object_dtype_isnan(X[i, cat_mask]) |
+                     _object_dtype_isnan(Y[j_start:, cat_mask]))
+
+        # Calculates the Manhattan distances for numerical columns
+        num_dists = abs(X[i, num_mask].astype(np.float32) -
+                        Y[j_start:, num_mask].astype(np.float32)) / ranges
+        # Calculates the number of non missing columns
+        n_missing = X.shape[1] - (np.isnan(cat_dists).sum(axis=1) +
+                                  np.isnan(num_dists).sum(axis=1))
+        # Gets the final results
+        results = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
+        results /= n_missing
+
+        D[i, j_start:] = results
+        if X is Y:
+            D[i:, j_start] = results
+
+    return D
+
+
+def _detect_categorical_features(X, categorical_features=None):
+    """Identifies the numerical and non-numerical (categorical) columns
+    of an array.
+
+    Parameters
+    ----------
+    X : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    categorical_features : array-like, optional, shape (n_features)
+        Indicates with True/False whether a column is a categorical attribute.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+        If the categorical_features array is None, they will be automatically
```

Comment:
This needs to be specified in a docstring of a public function, not a private one. And it needs to give more detail about the automatic detection. As it is, I'm not comfortable about this automatic detection stuff, unless the column contains strings or pd.Categorical.